platform/linux-generic/odp_ipsec.c

```diff
@@ -565,18 +593,40 @@ static int ipsec_in_esp(odp_packet_t *pkt,
 	param->auth_iv_ptr = state->iv;
 
 	state->esp.aad.spi = esp.spi;
-	state->esp.aad.seq_no = esp.seq_no;
 	state->in.seq_no = odp_be_to_cpu_32(esp.seq_no);
+	if (ipsec_sa->esn)
+		ipsec_sa->in.seq64 = ipsec_compute_esn(ipsec_sa, state->in.seq_no);
+
+	/* AAD with ESN */
+	if (ipsec_sa->esn && (!ipsec_sa->seqh_len))
+		state->esp.aad.seq_no64 = odp_cpu_to_be_64(ipsec_sa->in.seq64);
+	else if (!ipsec_sa->esn)
+		state->esp.aad.seq_no = esp.seq_no;
 
 	param->aad_ptr = (uint8_t *)&state->esp.aad;
 
+	/* Inbound ICV computation includes ESN higher 32 bits for individual algo's */
+	if (ipsec_sa->seqh_len) {
+		uint32_t inb_seqh = (ipsec_sa->in.seq64 >> 32);
+		uint32_t icv_offset =  odp_packet_len(*pkt) - ipsec_sa->icv_len;
+
+		if (odp_packet_add_data(pkt, icv_offset, ipsec_sa->seqh_len) < 0) {
```

Comment:
odp_packet_add_data() actually copies the whole packet (both in linux-gen and odp-dpdk). We do not want to do such a slow operation here.

Reply:
I am working on replacing this "odp_packet_add_data()" API  with some other odp_packet_* APIs, will push these changes as part of the next version of this PR.